File: matlab/test_allNORcircuits.m
clear
warning off
load('summary.mat')
load('maxgates.mat')
load('permuted_terms')
%nb_decimal = randi([0 65535])
QS_nb = -ones(65536,1);
nb_gates = zeros(65536,1);
QS_nb(1) = 0; QS_nb(65536) = 0;
parfor i = 2:65535
    subplot_nb = [1:6];
    nb_decimal = i - 1;
    all_perms_2 = unique(perms([1 2]),'rows');
    all_perms_3 = unique(perms([1 2 3]),'rows');
    all_perms_4 = unique(perms([1 2 3 4]),'rows');
    all_perms = {all_perms_2;all_perms_3;all_perms_4};
    remaining_terms = 1:4;
    remaining_terms(permuted_terms(dec_rep(nb_decimal+1,1)+1,:)==1) = [];

    nb_inputs = summary(nb_decimal+1,2);
    if nb_inputs == 1
        QS_nb(i) = 0;
        nb_gates(i) = 1;
    elseif nb_inputs == 2
        remaining_terms = remaining_terms(all_perms{1,1}(summary(dec_rep(nb_decimal+1,1)+1,4),:));
        nb_gates(i) = maxgates_2(find(inputs_2 == summary(dec_rep(nb_decimal+1,1)+1,3)));
        QS_nb(i) = print_graphs(summary(nb_decimal+1,2),find(inputs_2 == summary(dec_rep(nb_decimal+1,1)+1,3)),remaining_terms,subplot_nb,0);
    elseif nb_inputs == 3
        remaining_terms = remaining_terms(all_perms{2,1}(summary(dec_rep(nb_decimal+1,1)+1,4),:));
        nb_gates(i) = maxgates_3(find(inputs_3 == summary(dec_rep(nb_decimal+1,1)+1,3)));
        QS_nb(i) = print_graphs(summary(nb_decimal+1,2),find(inputs_3 == summary(dec_rep(nb_decimal+1,1)+1,3)),remaining_terms,subplot_nb,0);
    else
        remaining_terms = remaining_terms(all_perms{3,1}(summary(dec_rep(nb_decimal+1,1)+1,4),:));
        nb_gates(i) = maxgates_4(find(inputs_4 == summary(dec_rep(nb_decimal+1,1)+1,1)));
        QS_nb(i) = print_graphs(summary(nb_decimal+1,2),find(inputs_4 == summary(dec_rep(nb_decimal+1,1)+1,1)),remaining_terms,subplot_nb,0);
    end
end

save('QS_nb_N8.mat','QS_nb','nb_gates')

load ('nb_norgates.mat')
Nq71 = zeros(65536,1);
for i = 1:65536
   sum_i(i) = sum(nb_norgates{i,1} > 8); 
   if (sum_i(i) < 1)  || (QS_nb(i) == 1)
      Nq71(i) = 1; 
   end
   if (sum_i(i) == 1) && (length(nb_norgates{i,1})==1)
      Nq71(i) = 1; 
   end
end
sum(Nq71)/65536

load ('nb_norgates.mat')
Nq71 = zeros(65536,1);
for i = 1:65536
   sum_i(i) = sum(nb_norgates{i,1} > 7); 
   if (sum_i(i) < 4)  || (QS_nb(i) == 1) || (QS_nb(i) == 2)
      Nq71(i) = 1; 
   end
  if (sum_i(i) == 1) && (length(nb_norgates{i,1})==1)
      Nq71(i) = 1; 
   end
end
sum(Nq71)/65536
% 
figure('Renderer', 'painters', 'Position', [100 100 1000 405])
clf
subplot(1,2,1)
histogram(nb_gates,linspace(0.5,13.5,14))
[N,edges] = histcounts(nb_gates,linspace(0.5,14.5,15));
N = N/(sum(N));
N = cumsum(N);
xlabel('Minimum number of required NOR2 gates','FontWeight','Bold')
ylabel('Number of Boolean functions','FontWeight','Bold')
xticks(0:15)
grid on
subplot(1,2,2)
plot(1:14,N,'k','LineWidth',1.5)
hold on
scatter(1:14,N,'k')
xlabel('Minimum number of required NOR2 gates','FontWeight','Bold')
ylabel('Cumulative fraction of realizable Boolean functions','FontWeight','Bold')
grid on
xticks(0:15)
print('figure1.png','-dpng','-r300')

File: python/hybrid_logic/scripts/test_all_nor_circuits.py
"""Python port of ``test_allNORcircuits.m``.

This script iterates through every 4-input Boolean function (65,536 total),
looks up the canonical NOR2 realisations, and records the minimum gate usage
reported by the legacy MATLAB data.  It also generates the histogram and the
cumulative distribution plot found in the original workflow.

Usage:
    PYTHONPATH=python ./.venv/bin/python -m hybrid_logic.scripts.test_all_nor_circuits
"""
from __future__ import annotations

import argparse
from pathlib import Path
from typing import Tuple

import matplotlib.pyplot as plt
import numpy as np

from hybrid_logic.core.config import build_config_for_decimal
from hybrid_logic.core.print_graphs import build_tree_from_repo
from hybrid_logic.data.loader import HybridDataError, load_repository


def _compute_gate_statistics(limit: int | None = None) -> Tuple[np.ndarray, np.ndarray]:
    repo = load_repository()
    summary = repo.summary.summary
    qs_nb = np.full(65536, -1, dtype=np.int8)
    nb_gates = np.zeros(65536, dtype=np.int8)

    qs_nb[0] = 0
    qs_nb[-1] = 0

    iterator = range(1, 65535 if limit is None else min(limit, 65535))

    for i in iterator:
        nb_decimal = i
        nb_inputs = int(summary[nb_decimal, 1])
        if nb_inputs <= 1:
            qs_nb[nb_decimal] = 0
            nb_gates[nb_decimal] = 1 if nb_inputs == 1 else 0
            continue
        try:
            cfg = build_config_for_decimal(repo, nb_decimal)
        except HybridDataError:
            qs_nb[nb_decimal] = 0
            nb_gates[nb_decimal] = 0
            continue
        nb_gates[nb_decimal] = int(repo.tables.max_gates[nb_inputs][cfg.index])
        tree = build_tree_from_repo(repo, cfg)
        qs_nb[nb_decimal] = tree.nb_qs

    return qs_nb, nb_gates


def _plot_histogram(nb_gates: np.ndarray, output: Path) -> None:
    fig, axes = plt.subplots(1, 2, figsize=(10, 4), dpi=150)

    bins = np.linspace(0.5, 13.5, 14)
    axes[0].hist(nb_gates, bins=bins, edgecolor="black")
    axes[0].set_xlabel("Minimum number of required NOR2 gates", fontweight="bold")
    axes[0].set_ylabel("Number of Boolean functions", fontweight="bold")
    axes[0].grid(True, linestyle=":")
    axes[0].set_xticks(range(0, 15))

    counts, _ = np.histogram(nb_gates, bins=np.linspace(0.5, 14.5, 15))
    cumulative = np.cumsum(counts) / counts.sum()
    x = np.arange(1, 15)
    axes[1].plot(x, cumulative[:14], color="black", linewidth=1.5)
    axes[1].scatter(x, cumulative[:14], color="black")
    axes[1].set_xlabel("Minimum number of required NOR2 gates", fontweight="bold")
    axes[1].set_ylabel("Cumulative fraction of realisable Boolean functions", fontweight="bold")
    axes[1].grid(True, linestyle=":")
    axes[1].set_xticks(range(0, 15))

    fig.tight_layout()
    fig.savefig(output)
    plt.close(fig)


def main() -> None:
    parser = argparse.ArgumentParser(description="Analyse NOR2 gate counts for all Boolean functions.")
    parser.add_argument("--limit", type=int, default=None, help="Optional limit for debugging.")
    parser.add_argument(
        "--output",
        type=Path,
        default=Path("figure1.png"),
        help="Path to save the histogram figure (default: figure1.png).",
    )
    args = parser.parse_args()

    qs_nb, nb_gates = _compute_gate_statistics(limit=args.limit)
    _plot_histogram(nb_gates, args.output)
    print(f"Saved histogram to {args.output}")


if __name__ == "__main__":
    main()
